cpp/src/lists/drop_list_duplicates.cu

```diff
@@ -0,0 +1,288 @@
+  auto const device_input_table =
+    cudf::table_device_view::create(table_view{{all_lists_entries}}, stream);
+  auto const comp = row_equality_comparator<has_nulls>(
+    *device_input_table, *device_input_table, nulls_equal == null_equality::EQUAL);
+
+  auto const num_entries = all_lists_entries.size();
+  // Allocate memory to store the indices of the unique entries
+  auto const unique_indices = cudf::make_numeric_column(
+    entries_list_offsets.type(), num_entries, mask_state::UNALLOCATED, stream);
+  auto const unique_indices_begin = unique_indices->mutable_view().begin<size_type>();
+
+  auto const copy_end = thrust::unique_copy(
+    rmm::exec_policy(stream),
+    thrust::make_counting_iterator(0),
+    thrust::make_counting_iterator(num_entries),
+    unique_indices_begin,
+    [list_offsets = entries_list_offsets.begin<size_type>(), comp] __device__(
+      size_type i, size_type j) { return list_offsets[i] == list_offsets[j] && comp(i, j); });
+
+  // Collect unique entries and entry list offsets
+  auto const indices = cudf::detail::slice(
+    unique_indices->view(), 0, thrust::distance(unique_indices_begin, copy_end));
+  return cudf::detail::gather(table_view{{all_lists_entries, entries_list_offsets}},
+                              indices,
+                              cudf::out_of_bounds_policy::DONT_CHECK,
+                              cudf::detail::negative_index_policy::NOT_ALLOWED,
+                              stream,
+                              mr)
+    ->release();
+}
+
+/**
+ * @brief Generate a 0-based offset column for a lists column
+ *
+ * Given a lists_column_view, which may have a non-zero offset, generate a new column containing
+ * 0-based list offsets. This is done by subtracting each of the input list offset by the first
+ * offset.
+ *
+ * @code{.pseudo}
+ * Given a list column having offsets = { 3, 7, 9, 13 },
+ * then output_offsets = { 0, 4, 6, 10 }
+ * @endcode
+ *
+ * @param lists_column The input lists column
+ * @param stream       CUDA stream used for device memory operations and kernel launches
+ * @param mr           Device resource used to allocate memory
+ *
+ * @return A column containing 0-based list offsets
+ */
+std::unique_ptr<column> generate_clean_offsets(lists_column_view const& lists_column,
+                                               rmm::cuda_stream_view stream,
+                                               rmm::mr::device_memory_resource* mr)
+{
+  auto const list_offsets =
+    cudf::detail::slice(lists_column.offsets(),
+                        {lists_column.offset(), lists_column.offset() + lists_column.size() + 1},
+                        stream)
+      .front();
+  auto output_offsets = make_numeric_column(
+    list_offsets.type(), lists_column.size() + 1, mask_state::UNALLOCATED, stream, mr);
+  thrust::transform(
+    rmm::exec_policy(stream),
+    list_offsets.begin<size_type>(),
+    list_offsets.end<size_type>(),
```

Comment:
```suggestion
    list_offsets.begin<lists_column_view::offset_type>(),
    list_offsets.end<lists_column_view::offset_type>(),
```
Use `lists_column_view::offset_type` instead of `size_type` for type offset type.